pravega/zookeeper-operator
# Issue 66: Incorrect cluster membership

Signed-off-by: pbelgundi <prajakta.belgundi@emc.com>
### Change log description
Zookeeper dynamic configuration parameters (ex: servers in cluster) can change at run-time, and when that happens, the changed configuration needs to be persisted to disk for future reference. However, in the current zookeeper cluster, "/conf" directory is read-only (source for a config-map), due to which new dynamic configurations could not be persisted, giving rise to inconsistencies in cluster state. 
This PR also fixes the following issues with Readiness Probe :
1. Always failing on servers that start as "observers" because of `exit 1` added in block that upgrades observers to participants.
2. Post deletion and restart Server-0 remains as "observer" and never changes to "participant" as this server was ignored by the readiness probe script.
3. Added 'nslookup' to make sure upgrade is not attempted when service is not yet available ( for the very 1st probe on server 0, that has not yet started)

### Purpose of the change
Fixes #66

### What the code does
This fix changes the configuration location for zookeeper from "/conf" to "/data/conf" which is a writable folder, so new dynamic configurations can get persisted. All necessary files, including static configuration file "zoo.cfg" and copied into the new conf location, before startup.
Updated the 'zk-operator' image repository location to correctly point to "pravega/zookeeper-operator". 


### How to verify it
After this fix, did not see this error in zk logs: "Error writing next dynamic config file to disk".
Also cluster was in a consistent state, with exactly one node being "leader" and other being "followers".
Checked by executing "zkCli.sh status" on each server.

Also saw multiple dynamic configuration files being, as new severs were added to the cluster. The last one has the correct complete configuration and "/data/conf/zoo.cfg" points to it : 
```
/zookeeper-3.5.4-beta # ls -lt /data/conf/
-rw-r--r--    1 root     root           235 Jun 19 03:01 zoo.cfg
-rw-r--r--    1 root     root           484 Jun 19 03:01 zoo.cfg.dynamic.100000065
-rw-r--r--    1 root     root           387 Jun 19 03:01 zoo.cfg.dynamic.100000062
-rw-r--r--    1 root     root           481 Jun 19 03:00 zoo.cfg.dynamic.100000052
-rw-r--r--    1 root     root           387 Jun 19 03:00 zoo.cfg.dynamic.100000039
-rw-r--r--    1 root     root           290 Jun 19 03:00 zoo.cfg.dynamic.100000036
-rw-r--r--    1 root     root           384 Jun 19 03:00 zoo.cfg.dynamic.10000002c

/zookeeper-3.5.4-beta # cat /data/conf/zoo.cfg
initLimit=10
. . .
dynamicConfigFile=/data/conf/zoo.cfg.dynamic.100000065
```

On pod deletion ( 2 at a time) and restart, again new dynamic configuration files get generated on all servers, showing the change in cluster state:
```
/zookeeper-3.5.4-beta # ls -lt /data/conf/
total 56
-rw-r--r--    1 root     root           235 Jun 19 03:05 zoo.cfg
-rw-r--r--    1 root     root           484 Jun 19 03:05 zoo.cfg.dynamic.10000012a
-rw-r--r--    1 root     root           387 Jun 19 03:05 zoo.cfg.dynamic.100000127
-rw-r--r--    1 root     root           481 Jun 19 03:04 zoo.cfg.dynamic.100000111
-rw-r--r--    1 root     root           387 Jun 19 03:03 zoo.cfg.dynamic.1000000e5
-rw-r--r--    1 root     root           484 Jun 19 03:01 zoo.cfg.dynamic.100000065
-rw-r--r--    1 root     root           387 Jun 19 03:01 zoo.cfg.dynamic.100000062
-rw-r--r--    1 root     root           481 Jun 19 03:00 zoo.cfg.dynamic.100000052
-rw-r--r--    1 root     root           387 Jun 19 03:00 zoo.cfg.dynamic.100000039
-rw-r--r--    1 root     root           290 Jun 19 03:00 zoo.cfg.dynamic.100000036
-rw-r--r--    1 root     root           384 Jun 19 03:00 zoo.cfg.dynamic.10000002c

/zookeeper-3.5.4-beta # cat /data/conf/zoo.cfg
initLimit=10
. . .
dynamicConfigFile=/data/conf/zoo.cfg.dynamic.10000012a
```

All unit tests executed successfully.

Tested this fix on GKE as well as PKS clusters.
Deleted servers restart successfully and join the zk ensemble.
Server-0 which earlier used to stay in "obsever" state post restart, now  properly  migrates to "participant" after 1st readiness probe.
No readiness probe failures for servers starting as "observers"
